infra/src/main/kotlin/fi/fta/geoviite/infra/geometry/GeometryService.kt

```diff
@@ -301,6 +303,191 @@ class GeometryService @Autowired constructor(
             trackNumbersMatch(header, trackNumberIds) && freeTextMatches(header, searchTerms)
         }
     }
+
+    private fun collectSegmentSources(alignment: LayoutAlignment): List<SegmentSource> {
+        val planAlignmentIdToPlans: MutableMap<IntId<GeometryAlignment>, RowVersion<GeometryPlan>> = mutableMapOf()
+
+        return alignment.segments.map { segment ->
+            val sourceId = segment.sourceId
+            if (sourceId == null || sourceId !is IndexedId || segment.sourceStart == null) {
+                SegmentSource(null, null, null)
+            } else {
+                val planAlignmentId = IntId<GeometryAlignment>(sourceId.parentId)
+                val planVersion =
+                    planAlignmentIdToPlans.computeIfAbsent(planAlignmentId, geometryDao::fetchAlignmentPlanVersion)
+                val plan = geometryDao.fetchPlan(planVersion)
+                val planAlignment = plan.alignments.find { alignment -> alignment.id == planAlignmentId }
+                    ?: return@map SegmentSource(null, null, plan)
+                val planProfile = planAlignment.profile ?: return@map SegmentSource(null, null, plan)
```

Comment:
Tarvitaanko näitä returneja?
val planAlignment = plan.alignments.find(...)
val planProfile = planAlignment?.profile
val planElement = plantAligment?.elements?.get(...)
SegmentSource(...)

Reply:
:+1: parempi ilman aikaisia returneja, erityisesti kotlin lambdojen kanssa